tilings/algorithms/gridded_perm_reduction.py

```diff
@@ -0,0 +1,332 @@
+            cleanreq = tuple(
+                gp for gp in self._minimize(requirement) if gp.avoids(*obstructions)
+            )
+            # If cleanreq is empty, then can not contain this requirement so
+            # the tiling is empty.
+            if not cleanreq:
+                return [tuple()]
+            res.append(cleanreq)
+        return res
+
+    @cssmethodtimer("GriddedPermReduction.remove_redundant_requirements")
+    def remove_redundant_requirements(
+        self, requirements: List[Tuple[GriddedPerm, ...]],
+    ) -> List[Tuple[GriddedPerm, ...]]:
+        """
+        Remove all redundant requirement lists.
+        It is redundant if either:
+            - all of gridded perms contain some gridded permutation in the
+            same other requirement.
+            - all of the factors of the gridded permutations in the requirement
+            are contained in some other gridded permutation from the same other
+            requirement. (Is checking just this enough?)
+        """
+        idx_to_remove: Set[int] = set()
+        for i, requirement in enumerate(requirements):
+            for j, other_requirement in enumerate(requirements):
+                if i != j and j not in idx_to_remove:
+                    if all(
+                        self._griddedperm_implied_by_requirement(gp, other_requirement)
+                        for gp in requirement
+                    ):
+                        idx_to_remove.add(i)
+
+        for i, requirement in enumerate(requirements):
+            if i in idx_to_remove:
+                continue
+            factored_requirement = [r.factors() for r in requirement]
+            # if every factor of every requirement in a list is implied by
+            # another requirement then we can remove this requirement list
+            for factors in factored_requirement:
+                if all(
+                    self._griddedperm_implied_by_some_requirement(
+                        factor,
+                        (
+                            other_requirement
+                            for j, other_requirement in enumerate(requirements)
+                            if i != j and j not in idx_to_remove
+                        ),
+                    )
+                    for factor in factors
+                ):
+                    idx_to_remove.add(i)
+                    break
+        return sorted(
+            requirement
+            for idx, requirement in enumerate(requirements)
+            if idx not in idx_to_remove
+        )
+
+    @cssmethodtimer("GriddedPermReduction._griddedperm_implied_by_requirement")
+    def _griddedperm_implied_by_requirement(
+        self, griddedperm: GriddedPerm, requirement: Iterable[GriddedPerm]
+    ):
+        """
+        Return True, if the containment of the gridded perm is implied by
+        the containment of the requirement.
+        """
+        return all(griddedperm in gp for gp in requirement)
+
```

Comment:
# detect if there exists a requirement such that every component in that requirement
    # contains griddedperm